pkg/controller/cluster/cluster_controller.go

```diff
@@ -123,6 +124,36 @@ func (r *ReconcileCluster) Reconcile(request reconcile.Request) (reconcile.Resul
 			return reconcile.Result{}, nil
 		}
 
+		children, err := r.listChildren(context.Background(), cluster)
+		if err != nil {
+			klog.Infof("Failed to list dependent objects of cluster %s/%s: %v", cluster.ObjectMeta.Namespace, cluster.ObjectMeta.Name, err)
+			return reconcile.Result{}, err
+		}
+
+		if len(children) > 0 {
+			klog.Infof("deleting cluster %s: %d children still exist, will requeue", name, len(children))
+			for _, child := range children {
+
+				accessor, err := meta.Accessor(child)
+				if err != nil {
+					return reconcile.Result{}, errors.Wrapf(err, "couldn't create accessor for %T", child)
+				}
+
+				if accessor.GetDeletionTimestamp() != nil {
+					continue
+				}
+
+				gvk := child.GetObjectKind().GroupVersionKind().String()
+
+				klog.V(4).Infof("deleting cluster %s: Deleting %s %s", name, gvk, accessor.GetName())
+				if err := r.Delete(context.Background(), child, client.PropagationPolicy(metav1.DeletePropagationForeground)); err != nil {
+					return reconcile.Result{}, errors.Wrapf(err, "deleting cluster %s: failed to delete %s %s", name, gvk, accessor.GetName())
+				}
+			}
+
+			return reconcile.Result{Requeue: true}, nil
```

Comment:
Do we want to requeue immediately, or can we wait 10-20 seconds?

Reply:
Are we worried about overloading something? I figured it would go to the end of the queue so it wouldn't take up space on a busy cluster, and be more responsive on an empty one. Open to discussion though

Reply:
I figured Machines for example in AWS can take a while to get deleted, I don't have a strong opinion, but seems an immediate requeue might be putting pressure on the API server for little gain